stan/math/rev/mat/fun/cholesky_decompose.hpp

```diff
@@ -223,6 +233,133 @@ class cholesky_scalar : public vari {
     }
   }
 };
+#ifdef STAN_OPENCL
+class cholesky_gpu : public vari {
+ public:
+  int M_;
+  vari** variRefA_;
+  vari** variRefL_;
+
+  /**
+   * Constructor for GPU cholesky function.
+   *
+   * Stores varis for A.  Instantiates and stores varis for L.
+   * Instantiates and stores dummy vari for upper triangular part of var
+   * result returned in cholesky_decompose function call
+   *
+   * variRefL aren't on the chainable autodiff stack, only used for storage
+   * and computation. Note that varis for L are constructed externally in
+   * cholesky_decompose.
+   *
+   *
+   * @param A matrix
+   * @param L_A matrix, cholesky factor of A
+   */
+  cholesky_gpu(const Eigen::Matrix<var, -1, -1>& A,
+               const Eigen::Matrix<double, -1, -1>& L_A)
+      : vari(0.0),
+        M_(A.rows()),
+        variRefA_(ChainableStack::instance().memalloc_.alloc_array<vari*>(
+            A.rows() * (A.rows() + 1) / 2)),
+        variRefL_(ChainableStack::instance().memalloc_.alloc_array<vari*>(
+            A.rows() * (A.rows() + 1) / 2)) {
+    size_t pos = 0;
+    for (size_type j = 0; j < M_; ++j) {
+      for (size_type i = j; i < M_; ++i) {
+        variRefA_[pos] = A.coeffRef(i, j).vi_;
```

Comment:
I feel like there's been a discussion about `coeffRef` somewhere, I'll try to look on discourse.

From the docs it looks like `coeffRef` skips an assertion? So we might use it because it's faster

https://eigen.tuxfamily.org/dox/classEigen_1_1DenseCoeffsBase_3_01Derived_00_01WriteAccessors_01_4.html#ae66b7d18b2a85f3139b703126974c900